docs/RFCS/20170920_sql_consistency_checks.md

```diff
@@ -0,0 +1,559 @@
+- Check for invalid column families
+
+SQL constraints are checks on the logical table (`CHECK`, `NOT NULL`,
+`UNIQUE`, `FOREIGN KEY`).
+
+The rest of the checks are on the physical table.
+- Check for indexes using [STORING][] have the correct data
+- Tables using [FAMILY][] have the data organized into families
+  correctly
+- Invalid data encodings
+- Non-canonical encodings (i.e. data round-trips)
+- [Composite encodings][] are valid
+- Key encoding correctly reflects the ordering of the decoded values
+- Secondary indexes have entries for all rows in the table
+- Dangling index references (i.e. missing row but index entry retained)
+
+All physical table checks will be implemented with a KV check. All
+logical table checks can be done with other processes, with the
+exception of the `NOT NULL` constraint, which can only be checked with a
+KV check.
+
+### KV checks (excluding indexes)
+
+KV checks will be implemented using distSQL. In order to do so we will
+create a new distSQL processor, `TableChecker`. `TableChecker` is
+similar to `TableReader`, but it differs by doing additional checks
+while scanning through KV pairs and returning a list of check failures
+intead of data rows. These check failure rows will then be collected and
+presented upon completion, as a new distSQL schema.
+
+During execution of the `CHECK` plan, we will construct and execute a
+distSQL physical plan, similar to how the `SchemaChanger` does vi a.
+`CreateBackfiller`. This plan will be made to consist of `TableChecker`
+processors processing all the key spans relevant to the check.
+
+The underlying processing that happens behind both `TableReader` and
+`TableChecker` is the `RowFetcher`. Inside the current `RowFetcher` the
+method `processKv` is responsible for encoding specific work, which is
+precisely where the checks will be added.
+
+Inside `processKV`, we can do the following checks in the key iterator:
+- Check composite encodings are correct and round-trip
+- Check key encodings correctly reflect ordering of decoded values
+- Check the KV pair is well-formed (everything is present, no extra
+  data, checksum matches)
+- For primary keys, verify that the column family 0 (sentinel) is
+  present
+
+While iterating through keys some may be interleaved KV pairs from other
+tables. These can be checked that they do refer to a table that should
+be present.
+
+Execution of KV checks can be throttled by providing `KVFetcher` with a
+batch size and then sleeping the between processing batches.
+
+### Index checking
+
+Secondary indexes need to be checked that no entries are missing for the
+row data. Index references may also point to an invalid row, i.e. they
+are dangling.
+
+Secondary indexes present a complexity that we don't have with other KV
+checks: we need to reference other KV pairs which may not exist on the
+key span we're searching.
+
+The approach for doing this check in the efficiently is not yet
+determined. See unresolved questions.
+
```

Comment:
I think this is the most important part of this RFC and I want you to write another RFC to implement that because it overlaps a lot with the work we want to eventually do with schema changes.

The idea would be to pick a timestamp to recompute an index. This will be a distsql query that returns the entire index at a timestamp (that's the common part with the future schema change index creation). The gateway node receiving the index would then push the computed index down to the nodes where the indexes are using another distsql request to check the index. The distsql request would read the index from disk at the same fixed timestamp and check the index against the computed index.